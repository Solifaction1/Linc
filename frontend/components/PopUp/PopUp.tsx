import React, { useRef, useEffect } from 'react'

import { defaultProps, propsConfig } from './popUp.config'
import styles from './popUp.module.scss'
import ActionButton from './ActionButton/ActionButton'

const PopUp: React.FC<propsConfig> = (props: propsConfig) => {
	const wrapperRef = useRef(null)

	useEffect(() => {
		const handleClickOutside = (e: MouseEvent) => {
			if (wrapperRef.current === e.target) {
				if (props.allowClickOut && props.closePopUp !== undefined)
					props.closePopUp()
			}
		}

		const handleKeyPress = (e: KeyboardEvent) => {
			if (e.key === 'Escape') {
				if (props.allowKeyPress && props.closePopUp !== undefined)
					props.closePopUp()
			}
		}

		document.addEventListener('mousedown', handleClickOutside)
		document.addEventListener('keydown', handleKeyPress)
		return () => {
			document.removeEventListener('mousedown', handleClickOutside)
			document.removeEventListener('keydown', handleKeyPress)
		}
	}, [])

	return (
		<>
			{props.backgroundBlur && (
				<div className={styles.backgroundBlur} ref={wrapperRef} />
			)}
			<div
				className={`${props.className} ${styles.popUp} ${
					props.open && styles.open
				}`}
			>
				{props.children}
				<ActionButton
					buttonType="close"
					className={styles.closeButton}
					handleClick={() => props.closePopUp && props.closePopUp()}
				/>
			</div>
		</>
	)
}

PopUp.defaultProps = defaultProps

export default PopUp
