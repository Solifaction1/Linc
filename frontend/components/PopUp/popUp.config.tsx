import React from "react";

type propsConfig = {
  open?: boolean;
  children?: React.ReactNode;
  backgroundBlur?: boolean;
  allowClickOut?: boolean;
  allowKeyPress?: boolean;
  className?: string;
  handleClick?: React.MouseEventHandler<HTMLButtonElement>;
  closePopUp?: () => void;
};

const defaultProps = {
  open: true,
  children: <h2>Hier komt een grote titel</h2>,
  backgroundBlur: true,
  allowClickOut: true,
  allowKeyPress: true,
  className: "",
  handleClick: (): void => console.log("Default function"),
  closePopUp: (): void => console.log("Default function"),
};

export type { propsConfig };
export { defaultProps };
