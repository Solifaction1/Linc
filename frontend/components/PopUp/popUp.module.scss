@import 'styles/vars';
@import 'styles/utilities';

.popUp {
    @extend %absolute-container;
    z-index: 2;
    display: none;

    width: 100%;
    max-width: $popup-container;
    min-height: 10rem;

    background-color: $color-white;
    box-shadow: $box-shadow;

  
    @media (max-width: $break-md) {
        max-width: calc(100vw - 2rem);
    }

    &.open {
        display: block;
    }

    .closeButton{
        position: absolute;
        top: 0;
        right: 0;
    }
}

.backgroundBlur {
    position: absolute;
    top: 0;
    left: 0;

    width: 100%;
    height: 100%;

    background: rgba(black, 0.2);
}
