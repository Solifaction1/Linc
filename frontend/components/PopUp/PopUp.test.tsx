import React from 'react'

import { render } from '@testing-library/react'
import '@testing-library/jest-dom'
import PopUp from './PopUp'

describe('PopUp', () => {
	it('renders a PopUp', () => {
		const { container, getByText } = render(<PopUp className="className" />)

		expect(container).toMatchSnapshot()

		expect(container.getElementsByClassName('backgroundBlur')[0]).toBeVisible()
		expect(
			container.getElementsByClassName('className popUp open')[0]
		).toBeVisible()

		expect(
			getByText('Hier komt een grote titel', { selector: 'h2' })
		).toBeVisible()

		const button = container.querySelector('button')
		expect(button).toBeVisible()
		expect(button?.getAttribute('class')).toContain('closeButton')
		expect(button?.getAttribute('class')).toContain('actionButton')
		expect(button?.getAttribute('class')).toContain('close')
		expect(button?.getAttribute('class')).toContain('md')
		expect(button?.getAttribute('type')).toEqual('button')
	})
})
