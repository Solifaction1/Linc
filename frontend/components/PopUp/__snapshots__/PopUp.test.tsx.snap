// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PopUp renders a PopUp 1`] = `
<div>
  <div
    class="backgroundBlur"
  />
  <div
    class="className popUp open"
  >
    <h2>
      Hier komt een grote titel
    </h2>
    <button
      class="
			closeButton 
			actionButton 
			close 
			md"
      type="button"
    >
      <div />
      <div />
    </button>
  </div>
</div>
`;
